p4-16/psa/PSA.mdk

```diff
@@ -58,38 +58,52 @@ interface ports.
 # Target Architecture Model
 
 
+As an analogy, the PSA is to the P4~16~ language as the C standard
+library is to the C programming language.  PSA defines a library of
+types, P4~16~ externs for frequently used constructs such as counters,
+meters, and registers, and a set of "packet paths" that enable you to
+write P4 programs that control the flow of packets in a packet switch
+that has multiple ports, e.g. dozens of Ethernet ports.  By following
```

Comment:
Are you equating the "packet paths" with the pipeline? I kind of see those a bit different. It is true that we said right in the beginning that: "P4 programs specify how the various programmable blocks of a target architecture are programmed and connected". However, it is really PSA that defines the structure of programmable blocks and how they are connected. Somehow we missed this in the different iterations.

I see that you expand on this below, but it is never quite clear that the architecture defines these blocks -- in P4 no less :)

Reply:
The packet paths are what we call packet paths later in the document, e.g. to/from ports, normal unicast/multicast packets through the packet buffer, resubmit, recirculate, and clone.  Those are fairly specific to PSA, and things that I suspect would look somewhat different if there were a "Portable NIC Architecture" document.  The PSA defines these packet paths, but mostly in English and pseudocode.

Reply:
Right, but also the role of PSA is to define the programmable blocks, and while we do discuss these later, it seems that we lead with defining most of the other things.